Product Gallery: Add animation when large image changes

gigitux opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

> [!WARNING]  
> This PR is blocked by woocommerce/woocommerce-blocks#11111 

> [!NOTE]  
> I created a follow-up issue to add E2E tests woocommerce/woocommerce#42192


## What

Fixes woocommerce/woocommerce-blocks#10870. Currently, there isn't any animation when the large image changes.

## Why

This PR implements the animation when the user changes the large image with:
- slider
- thumbnails
- arrows button (in progress - https://github.com/woocommerce/woocommerce-blocks/pull/10938/)

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Edit the Single Product Template.
2. Add the Product Gallery block.
3. Save the template.
4. Visit a product that has multiple images.
5. Change the large image by clicking on the thumbnails and with the slider.
6. Ensure that the animation works correctly.


* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast




<!-- Any screenshots of UI changes will be helpful to include here. Leave blank if not applicable. -->

| Before | After |
| ------ | ----- |
|<video src=https://github.com/woocommerce/woocommerce-blocks/assets/4463174/d2b8ab93-9452-4c61-86d6-9851f4af667a />|<video src=https://github.com/woocommerce/woocommerce-blocks/assets/4463174/1016bd13-df97-4e88-b2e2-1766e749277c />|

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [x] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Add suggested changelog entry here.

gigitux committed:
Add slide animation

albarin committed:
Remove placeholder and pagination (#11145)

* Add titles to patterns and set the aligment to Wide

* Replace product query patterns with product collection ones

* Remove pagination and no results query from product query patterns

* Add aspect ratio to the product image attributes

* Add portrait aspect ratio to product X column and product gallery patterns

gigitux committed:
improve animation

gigitux committed:
Merge branch 'trunk' into fix/animation-dialog

gigitux committed:
improve naming

gigitux committed:
Merge branch 'fix/animation-dialog' of github.com:woocommerce/woocommerce-blocks into fix/animation-dialog

gigitux committed:
fix regression

roykho commented:
Bumping this PR to next release.

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-blocks into fix/animation-dialog

danieldudzic commented:
> This PR is blocked by https://github.com/woocommerce/woocommerce-blocks/pull/11111

Do you mean by this, the following regression with the `Next/Prev Buttons being inside the image`?
![XZF9vBZCYr](https://github.com/woocommerce/woocommerce-blocks/assets/905781/6a53feef-c44f-4e4b-980a-0a194a04de90)

It seems to be caused by https://github.com/woocommerce/woocommerce-blocks/pull/11111

danieldudzic commented:
Good job on this! The animation overall looks very convincing, but I have noticed one thing.

The animation seems to be a bit fragile - perhaps this is due to the Zoom effect, but after a few swipes the display of the Large Image breaks a bit and the neighbor image leaks into the Large Image:
![FaNlw6RaUZ](https://github.com/woocommerce/woocommerce-blocks/assets/905781/1f4aaad3-c669-44fc-ad1d-c71026d06ff0)

danieldudzic commented on assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/frontend.tsx:
```diff
@@ -78,5 +91,59 @@ interactivityStore(
+				scrollInto: ( store: Store ) => {
+					// Scroll to the selected image with a smooth animation.
+					if (
+						store.selectors?.woocommerce?.isSelected( store ) &&
```
`store.selectors?.woocommerce?.isSelected( store )` seems to be present in all the checks.

Can we perhaps handle it in a top-level condition?

gigitux replied:
Good point! I fixed it with [8fe7233](https://github.com/woocommerce/woocommerce-blocks/pull/11113/commits/8fe72333fde771e7bca133d7cbf1b6e34ab481bb).

danieldudzic commented on assets/js/blocks/product-gallery/inner-blocks/product-gallery-large-image/frontend.tsx:
```diff
@@ -78,5 +91,59 @@ interactivityStore(
+						store.context.woocommerce.isDialogOpen !==
+							isDialogStatusChanged &&
+						store.selectors?.woocommerce?.isSelected( store ) &&
+						store.ref.tagName === 'IMG' &&
```
This is me being very nit-picky; I believe we should avoid targeting the `tagName` as it is too broad.

A bit of a different example, but I believe it illustrates the point. I have changed it to `classList.contains` in https://github.com/woocommerce/woocommerce-blocks/pull/11087 
so that the View All thumbnails overlay (which is not an IMG) could trigger the opening of the Dialog too.

gigitux replied:
Thanks for the suggestion, but I don't think that it is an approach that we should follow (I will add a specific comment in #11087).

I realized that the check on the tag IMG isn't necessary. So, I just removed it :)

danieldudzic reviewed (changes requested):
Very impressive work @gigitux 🙇 

I'm setting this as `Request changes` as the animation needs to be refined to make sure it always displays the Large Image correctly, and the CSS leak fix regression needs to be addressed as well.

gigitux committed:
fix css

gigitux committed:
improve code style

gigitux committed:
remove check on tag image

gigitux commented:
> Very impressive work @gigitux 🙇
> 
> I'm setting this as `Request changes` as the animation needs to be refined to make sure it always displays the Large Image correctly, and the CSS leak fix regression needs to be addressed as well.

Thanks for the review! Nice catch! With [69daeae](https://github.com/woocommerce/woocommerce-blocks/pull/11113/commits/69daeaeb81981e091170f87eb1fa014357c952e1), I tried to fix the CSS. Please double-check it and let me know if you think that it could introduce some regression.

This is the result:


https://github.com/woocommerce/woocommerce-blocks/assets/4463174/9edff895-bdb9-44e8-a44c-32b46441d8fb

gigitux committed:
Merge branch 'trunk' into fix/animation-dialog

danieldudzic commented:
> > Very impressive work @gigitux 🙇
> > I'm setting this as `Request changes` as the animation needs to be refined to make sure it always displays the Large Image correctly, and the CSS leak fix regression needs to be addressed as well.
> 
> Thanks for the review! Nice catch! With https://github.com/woocommerce/woocommerce-blocks/pull/11113/commits/69daeaeb81981e091170f87eb1fa014357c952e1, I tried to fix the CSS. Please double-check it and let me know if you think that it could introduce some regression.
> 
> This is the result:
> Screen.Capture.Oct.11.mov

It seems to be working mostly fine with the "outside" button setting (the height of the Large Image is static though):

<img width="1069" alt="V-Neck_T-Shirt_–_newproductgallery-4" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/f354b3ce-3360-4db8-8be9-c372bf3c2d0b">

It is still broken with the "inside" button setting:
![QQsMiULyLj](https://github.com/woocommerce/woocommerce-blocks/assets/905781/3845542d-af1f-4f05-b617-85f0f6a52bdd)

gigitux commented:
I don't think it is a regression. The current behavior in trunk is broken as well:


https://github.com/woocommerce/woocommerce-blocks/assets/4463174/b28deeaa-c0e0-480e-b7bb-cb91ace6d6dc

I think that it should be a separate discussion about how to handle different sizes of the images:
<img width="274" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/4463174/537c143e-0ef2-4ea2-b4c6-413659ba1340">

We should implement the "crop images to fit" setting. So, all the images will fit in the square (even if stretched). When it is disabled, I think that the behavior in the video that you shared is correct.

danieldudzic commented:
1. There's a difference in the height of the Large Image:

| trunk | fix/animation-dialog |
| ------ | ----- |
|<img width="1256" alt="V-Neck_T-Shirt_–_newproductgallery-5" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/4414ba5c-b3a4-4c43-99c2-abea6fd783af">|<img width="1027" alt="V-Neck_T-Shirt_–_newproductgallery-6" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/6a21154c-8740-4a01-b9d7-73387dff79d3">|

2. I agree that image cropping is a separate concern, but the sliding should still handle correctly both scenarios (and not set a fixed height):
- Images has the same height (square)
- Images differ in height

3. Regarding the issue with the sliding/image overflow:
![274435412-3845542d-af1f-4f05-b617-85f0f6a52bdd](https://github.com/woocommerce/woocommerce-blocks/assets/905781/c7b78410-8ad2-41ef-8863-1a46f6e757bc)

I can't replicate it anymore. ✅

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-blocks into fix/animation-dialog

gigitux commented:
> There's a difference in the height of the Large Image:


> I agree that image cropping is a separate concern, but the sliding should still handle correctly both scenarios (and not set a fixed height):

Both potential issues are correlated. With the current implementation, the row of images is handled by Flexbox. The highest image sets the height. I think that this approach is the right one because the position of the arrow has to be fixed and not be dependent by the width/height of the image like currently happens on trunk:

https://github.com/woocommerce/woocommerce-blocks/assets/4463174/b28deeaa-c0e0-480e-b7bb-cb91ace6d6dc

I think that when "image cropping" is enabled, we can force all the images to have the maximum width and height available. When it is disabled, could we center the image? What do you think?

https://github.com/woocommerce/woocommerce-blocks/assets/4463174/3751f7d0-cacc-4476-9a49-b9d26c436052

gigitux committed:
Merge branch 'trunk' into fix/animation-dialog

danieldudzic commented:
> Both potential issues are correlated. With the current implementation, the row of images is handled by Flexbox. The highest image sets the height. I think that this approach is the right one because the position of the arrow has to be fixed and not be dependent by the width/height of the image like currently happens on trunk:

1. I believe the common pattern is that the "slider" elements generally adapt to the image size (height) when cropping is disabled. We should get some design input on this.

2. I don't have clarity regarding how cropping exactly should work, but yes, perhaps cropping everything to fit nicely in a square similar to the original Product Gallery would probably make most sense.

gigitux commented:
Hey @jarekmorawski, could you give us feedback on this change, please? You can take a look to [this comment](https://github.com/woocommerce/woocommerce-blocks/pull/11113#issuecomment-1764026705).

Personally, I think that the behavior in the first video is wrong, and we should stick with the second behavior (maybe align the image to the center).

Instead, when the user enables the setting "Crop image to fit", all the images will have the gallery's height and width. Does it make sense?

gigitux committed:
Merge branch 'trunk' into fix/animation-dialog

gigitux committed:
Merge branch 'trunk' into fix/animation-dialog

gigitux committed:
Merge branch 'trunk' into fix/animation-dialog

gigitux committed:
Merge branch 'trunk' into fix/animation-dialog

jarekmorawski commented:
> I think that when "image cropping" is enabled, we can force all the images to have the maximum width and height available. When it is disabled, could we center the image? What do you think?

I agree. That's what the toggle does – if enabled, all images will be made into squares and aligned to the center (also during crop; most images have a central focal point, so we should always crop out the edges).

> Personally, I think that the behavior in the first video is wrong, and we should stick with the second behavior (maybe align the image to the center).

Agreed. The behavior in the first video feels a bit odd, but the worst part is that it changes the position of the arrow buttons, which isn't great. I think it'd make the most sense to align the images in the center no matter their aspect ratio.

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-blocks into fix/animation-dialog

gigitux committed:
Merge branch 'fix/animation-dialog' of github.com:woocommerce/woocommerce-blocks into fix/animation-dialog

gigitux committed:
align image

gigitux committed:
fix crash when zoom is disabled

gigitux committed:
fix E2E tests

gigitux commented:
@jarekmorawski Thanks for the feedback!

@danieldudzic, according with the feedback, with [05a3eb6](https://github.com/woocommerce/woocommerce-blocks/pull/11113/commits/05a3eb6f0e85aacafcb1b5b0786cb88227f07d14) I added the CSS to center the images. Also, I fixed an error when the zoom is disabled (18d416e) and fixed some E2E tests given that the structure of the DOM is changed with this PR (b6d0088).

gigitux committed:
improve CSS

danieldudzic reviewed (approved):
Superb work, @gigitux!

Both the animation and zoom now work as expected!
